voucher/views.py

```diff
@@ -1,3 +1,180 @@
+from django.contrib.auth.mixins import LoginRequiredMixin
+from django.db.models import Q
+from django.shortcuts import render, redirect, get_object_or_404
+from django.urls import reverse
+from django.views.generic import FormView
+
+from courses.models import CourseRun
+from ecommerce.models import CouponRedemption, CouponVersion, Product
+from ecommerce.api import get_valid_coupon_versions
+from mitxpro.views import get_js_settings_context
+from voucher.forms import UploadVoucherForm, CourseRunSelectForm
+from voucher.models import Voucher
+
+
+class UploadVoucherFormView(LoginRequiredMixin, FormView):
+    """
+    UploadVoucherFormView displays the voucher upload form and handles its submission
+    """
+
+    template_name = "upload.html"
+    form_class = UploadVoucherForm
+
+    def form_valid(self, form):
+        """
+        If parsing is successful, route based on the voucher
+        """
+        return self.handle_voucher(form.cleaned_data["voucher"])
+
+    def form_invalid(self, form):
+        """
+        If parsing fails, route to email page, else return to form
+        """
+        if form.parse_error_string in form.errors["voucher"]:
+            return redirect(reverse("voucher:resubmit"))
+        return super().form_invalid(form)
+
+    def get_context_data(self, **kwargs):
+        return {
+            **super().get_context_data(**kwargs),
+            **get_js_settings_context(self.request),
+        }
+
+    def handle_voucher(self, values):
+        """
+        Get or create voucher and route based on current progress (i.e. coupon is redeemed / attached / unattached)
+        :param values: dictionary parsed from voucher PDF
+        """
+        user = self.request.user
+        # Check for an existing voucher
+        old_voucher = get_current_voucher(user)
+        # If a voucher exists, check if it is the same as the uploaded voucher
+        if old_voucher and all(
+            value in old_voucher.__dict__.items() for value in values.items()
+        ):
+            voucher = old_voucher
+            # Check if the voucher is already redeemed
+            if (
+                voucher.coupon
+                and CouponRedemption.objects.filter(
+                    coupon_version=CouponVersion.objects.filter(
+                        coupon=voucher.coupon
+                    ).last()
+                ).first()
+            ):
+                return redirect("voucher:redeemed")
+        else:
+            Voucher.objects.create(**values, user=user)
+
+        return redirect("voucher:enroll")
+
+
+class CourseRunSelectFormView(LoginRequiredMixin, FormView):
+    """
+    VoucherListView displays all active vouchers for the user along with all potential course matches
+    """
+
+    template_name = "enroll.html"
+    form_class = CourseRunSelectForm
+
+    def get(self, request, *args, **kwargs):
+        voucher = get_current_voucher(self.request.user)
+        if voucher is None:
```

Comment:
Maybe we should check here to see if the voucher coupon has already been redeemed, as is done in the `UploadVoucherFormView` above?  

I manually completed an order for the voucher product and redeemed the coupon after submitting this form.  Then I returned to `../boeing/enroll` and was still able to submit here again.

Reply:
Since a handle voucher redirects to the enroll page, I can remove the check from there and only checking when accessing enroll.